src/modules/battery_resistance_est/battery_resistance_est.hpp

```diff
@@ -0,0 +1,139 @@
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAfieldE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file BatteryResistanceEstimator.hpp
+ *
+ * Estimator for the battery internal resistance parameter to run online using only the current and voltage data​.
+ *
+ * @author Mohamad Akkawi	<akkawi@protonmail.com>
+ *
+ * Implementation based on 'Online estimation of internal resistance and open-circuit voltage of lithium-ion
+ * batteries in electric vehicles' by Yi-Hsien Chiang , Wu-Yang Sean, Jia-Cheng Ke
+ *
+ */
+
+#pragma once
+
+#include <px4_platform_common/module.h>
+#include <px4_platform_common/module_params.h>
+#include <px4_platform_common/px4_work_queue/ScheduledWorkItem.hpp>
+
+#include <uORB/Publication.hpp>
+#include <uORB/Subscription.hpp>
+#include <uORB/SubscriptionCallback.hpp>
+
+#include <uORB/topics/battery_status.h>
+#include <uORB/topics/internal_resistance.h>
+
+#include <matrix/math.hpp>
+
+#include <uORB/topics/vehicle_status.h>
+
+using namespace matrix;
+
+class InternalRes : public ModuleBase<InternalRes>, public ModuleParams, public px4::WorkItem
+{
+public:
+	InternalRes();
+	~InternalRes() {};
+
+	static int task_spawn(int argc, char *argv[]);
+
+	static int custom_command(int argc, char *argv[]);
+
+	static int print_usage(const char *reason = nullptr);
+
+	bool init();
+
+	int print_status() override;
+
+	void update_internal_resistance(const float voltage_estimation_error, const Vector<float, 4> esm_params_est);
+
+	Vector<float, 4> extract_ecm_parameters();
+
+	float predict_voltage(const float dt);
+
+private:
+	void Run() override;
+
+	internal_resistance_s _inter_res;
+	battery_status_s _battery_status;
+	vehicle_status_s _vehicle_status;
+
+	hrt_abstime _battery_time_prev;
+	hrt_abstime _battery_time;
+
+	hrt_abstime _last_param_update_time;
+
+	Vector<float, 4> _param_est;
+	Vector<float, 4> _adaptation_gain;
+	Vector<float, 4> _signal;
+	Vector<float, 4> _best_ecm_params_est;
+
+	float _lambda = 0.8f;
+	float _voltage_estimation;
+
+	float _current_filtered_a = 0.f;
+	float _voltage_filtered_v;
+	float _current_filtered_a_prev;
+
+	float _best_prediction_error;
+	bool _best_prediction_error_reset = true;
+
+//used to save estimated ecm params on disarm
+	bool _armed = false;
+	bool _was_armed = false;
+	bool _on_standby = false;
+	bool _param_bat_saved = false;
+
+	uORB::Subscription _vehicle_status_sub{ORB_ID(vehicle_status)};
+
+	uORB::SubscriptionCallbackWorkItem _battery_sub{this, ORB_ID(battery_status)};
+
+	uORB::Publication<internal_resistance_s> _internal_res_pub{ORB_ID(internal_resistance)};
+
+	DEFINE_PARAMETERS(
+		(ParamFloat<px4::params::BAT1_R_INTERNAL>) _param_bat1_r_internal,
+		(ParamFloat<px4::params::BAT1_V_CHARGED>) _param_bat1_v_charged,
+		(ParamInt<px4::params::BAT1_N_CELLS>) _param_bat1_n_cells,
```

Comment:
Note this doesn't necessarily scale properly on systems with multiple batteries as a particular instance of `battery_status` doesn't necessarily correspond to the same BATx parameters. Ignoring this initially is fine as long as this isn't going to run by default.

Reply:
The default for RIN_ENABLED is 0 
and in rcS of init.d and init.d-posix I have 
`	if param greater -s RIN_ENABLED 0
	then
		battery_resistance_est start
	fi` 

Let me know if there is something else I need to do to prevent it from running by default.

and thanks, yea I see the issue with scaling. If it's okay, lets ignore this for now and ill come back with another pr soon.